frontend/src/metabase/visualizations/echarts/cartesian/model/axis.ts

```diff
@@ -0,0 +1,251 @@
+
+  if (hasDifferentYAxisColTypes) {
+    return true;
+  }
+
+  const allMetricsColumnSettings = allMetricsColumns
+    .map(column => settings.column?.(column))
+    .filter(isNotNull);
+
+  const columnSettings = allMetricsColumnSettings.map(columnSettings =>
+    getLineAreaBarComparisonSettings(columnSettings),
+  );
+
+  const hasDifferentColumnSettings = columnSettings.some(s1 =>
+    columnSettings.some(s2 => !_.isEqual(s1, s2)),
+  );
+
+  if (hasDifferentColumnSettings) {
+    return true;
+  }
+
+  const yExtents = Object.values(seriesExtents);
+
+  const minRange = Math.min(...yExtents.map(extent => extent[1] - extent[0]));
+  const maxExtent = Math.max(...yExtents.map(extent => extent[1]));
+  const minExtent = Math.min(...yExtents.map(extent => extent[0]));
+  const chartRange = maxExtent - minExtent;
+
+  // Note (EmmadUsmani): When the series with the smallest range is less than 5%
+  // of the chart's total range, we split the y-axis so it doesn't look too small.
+  return minRange / chartRange <= 0.05;
+}
+
+type AxisSplit = [DataKey[], DataKey[]];
+
+const SPLIT_AXIS_UNSPLIT_COST = -100;
+const SPLIT_AXIS_COST_FACTOR = 2;
+const SPLIT_AXIS_MAX_DEPTH = 8;
+
+function generateSplits(
+  list: DataKey[],
+  left: DataKey[] = [],
+  right: DataKey[] = [],
+  depth = 0,
+): AxisSplit[] {
+  // NOTE: currently generates all permutations, some of which are equivalent
+  if (list.length === 0) {
+    return [[left, right]];
+  } else if (depth > SPLIT_AXIS_MAX_DEPTH) {
+    // If we reach our max depth, we need to ensure that any item that haven't been added either list are accounted for
+    return left.length < right.length
+      ? [[left.concat(list), right]]
+      : [[left, right.concat(list)]];
+  } else {
+    return [
+      ...generateSplits(
+        list.slice(1),
+        left.concat([list[0]]),
+        right,
+        depth + 1,
+      ),
+      ...generateSplits(
+        list.slice(1),
+        left,
+        right.concat([list[0]]),
+        depth + 1,
+      ),
+    ];
+  }
+}
+
+function axisCost(extents: Extent[], favorUnsplit = true) {
+  const axisExtent = d3.extent(extents.flatMap(e => e));
+  const axisRange = axisExtent[1] - axisExtent[0];
+  if (favorUnsplit && extents.length === 0) {
+    return SPLIT_AXIS_UNSPLIT_COST;
+  } else if (axisRange === 0) {
+    return 0;
+  } else {
+    return extents.reduce(
+      (sum, seriesExtent) =>
+        sum +
+        Math.pow(
+          axisRange / (seriesExtent[1] - seriesExtent[0]),
+          SPLIT_AXIS_COST_FACTOR,
+        ),
+      0,
+    );
+  }
+}
+
+export function computeSplit(
+  extents: SeriesExtents,
+  left: DataKey[] = [],
+  right: DataKey[] = [],
+): AxisSplit {
+  const unassigned: DataKey[] = Object.keys(extents).filter(
+    key => left.indexOf(key) < 0 && right.indexOf(key) < 0,
+  );
+
+  // if any are assigned to right we have decided to split so don't favor unsplit
+  const favorUnsplit = right.length > 0;
```

Comment:
What if there are keys assigned to the `right` array defined on line 214 in the parent function? Should we consider those?